File: docs/experiments/04-auto-compilation/auto_compilation.py
#!/usr/bin/env python3
"""Experiment 4 -- Auto-compilation: the compiler runs without Compile[].

Runs the same workloads as ``auto_compilation.m``.  See ``README.md``.

    python3 auto_compilation.py

WHAT THIS COLUMN IS.  The experiment's subject -- a CAS noticing that ordinary
user code is numeric and compiling it without being asked -- has no Python
counterpart, because SciPy's quadrature and ODE routines are already compiled
C and never saw a Python-level expression to optimise.  So these rows are not
"the same feature, three ways"; they are the standard library each system
would actually be measured against.

That makes two rows worth reading closely.  ``solve_ivp`` is a Python callback
per step, which is why the CAS beat it by a wide margin once they compile the
right-hand side.  ``CubicSpline`` is entirely compiled C and is correspondingly
hard to beat.
"""

import time

import numpy as np


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))

import math

from scipy import integrate, interpolate, optimize


def f(x):
    return math.sin(x) * math.exp(-x / 10.0) + math.sqrt(x + 1.0)


def fv(x):
    return np.sin(x) * np.exp(-x / 10.0) + np.sqrt(x + 1.0)


def lorenz(t, s):
    return [10.0 * (s[1] - s[0]),
            s[0] * (28.0 - s[2]) - s[1],
            s[0] * s[1] - (8.0 / 3.0) * s[2]]


def main():
    print("Experiment 4 -- auto-compilation")
    print("")
    print("-- numeric routines over a user function --")

    bench("quad(f, 0, 50)", lambda: integrate.quad(f, 0, 50))
    check("quad(f, 0, 50)", round(integrate.quad(f, 0, 50)[0], 6))

    ks = np.arange(1, 2001, dtype=float)
    bench("sum(f(k)/k^2, k = 1..2000)", lambda: (fv(ks) / ks ** 2).sum())
    bench("prod(1 + 1/k^2, k = 1..2000)", lambda: np.prod(1.0 + 1.0 / ks ** 2))
    bench("brentq(f(x) - 2)", lambda: optimize.brentq(lambda x: f(x) - 2, 0.1, 5))

    print("")
    print("-- elementwise and scan over a machine-number list --")
    lst = np.random.rand(10 ** 6)

    def nest_scalar(m):
        x = 0.31
        for _ in range(m):
            x = 3.5 * x * (1.0 - x)
        return x

    bench("nest 3.5 x (1-x), 10^6 (CPython scalar loop)",
          lambda: nest_scalar(10 ** 6), reps=1)
    bench("cumsum(sin(list))", lambda: np.cumsum(np.sin(lst)))
    bench("sin(list)**2 + 1", lambda: np.sin(lst) ** 2 + 1.0)
    bench("cumsum(list)", lambda: np.cumsum(lst))

    print("")
    print("-- an ODE and an interpolation --")
    bench("solve_ivp, Lorenz to t = 200",
          lambda: integrate.solve_ivp(lorenz, (0.0, 200.0), [1.0, 1.0, 1.0],
                                      rtol=1e-8, atol=1e-8), reps=1)

    ni = 10 ** 4
    xs = np.linspace(0.0, 100.0, ni)
    ifn = interpolate.CubicSpline(xs, np.sin(xs))
    pts = np.random.rand(ni) * 100.0
    bench("CubicSpline, 10^4 nodes, 10^4 evaluations", lambda: ifn(pts))
    check("CubicSpline at 3.3", round(float(ifn(3.3)), 6))


if __name__ == "__main__":
    main()

File: docs/experiments/04-auto-compilation/auto_compilation.m
(* ==========================================================================
   Experiment 4 -- Auto-compilation: the compiler runs without Compile[]
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file auto_compilation.m
       Mathematica   wolframscript -file auto_compilation.m
       Python        python3 auto_compilation.py

   WHAT IT MEASURES.  Ordinary user code -- no Compile[] anywhere -- that the
   system recognises as numeric and compiles on its own.  Two families:

     * the numeric BUILTINS, which evaluate a user function many times at
       machine precision (NIntegrate, Plot, FindRoot, NSum, NProduct, NDSolve)
     * the FUNCTIONAL heads, where the function is applied element by element
       (Nest, NestList, FoldList, Scan, Accumulate)

   In both cases the win is the same one experiment 1 measured, and the point
   of THIS experiment is that the user never asked for it.

   Every timing below can be compared against the un-compiled path by setting
   the environment variable MATHILDA_NO_AUTOCOMPILE=1 and re-running, which is
   how the README's before/after column was produced.
   ========================================================================== *)

SetAttributes[bench, HoldRest];

(* bench[label, expr] -- one untimed warm-up, then the MINIMUM of three timed
   runs.  The minimum, not the mean: we are measuring the cost of the work, and
   every source of noise on a loaded machine can only add. *)
bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];


Print["Experiment 4 -- auto-compilation"];
Print[""];
Print["-- numeric builtins over a user function --"];

f[x_] := Sin[x] Exp[-x/10.] + Sqrt[x + 1.];

bench["NIntegrate[f, {x, 0, 50}]",        NIntegrate[f[x], {x, 0, 50}]];
check["NIntegrate[f, {x, 0, 50}]",        Round[NIntegrate[f[x], {x, 0, 50}], 1.*^-6]];

bench["NSum[f[k]/k^2, {k, 1, 2000}]",     NSum[f[k]/k^2, {k, 1, 2000}]];
bench["NProduct[1 + 1/k^2, {k, 1, 2000}]", NProduct[1 + 1./k^2, {k, 1, 2000}]];
bench["FindRoot[f[x] == 2, {x, 1}]",      FindRoot[f[x] == 2, {x, 1}]];

Print[""];
Print["-- functional heads over a machine-number list --"];

lst = RandomReal[{0, 1}, 10^6];

bench["Nest[3.5 # (1-#) &, 0.31, 10^6]",   Nest[3.5 # (1. - #) &, 0.31, 10^6]];
bench["NestList[3.5 # (1-#) &, 0.31, 10^6]", NestList[3.5 # (1. - #) &, 0.31, 10^6]];
bench["FoldList[#1 + Sin[#2] &, 0., list]", FoldList[#1 + Sin[#2] &, 0., lst]];
bench["Scan[Sin[#]^2 + 1. &, list]",        Scan[Sin[#]^2 + 1. &, lst]];
bench["Accumulate[list]",                   Accumulate[lst]];
bench["NestWhileList[# + 1. &, 0., # < 10^6 &]",
      NestWhileList[# + 1. &, 0., # < 1.*^6 &]];

Print[""];
Print["-- an ODE and an interpolation, both of which evaluate a user body --"];

Clear[ly];
bench["NDSolve, Lorenz to t = 200",
      NDSolve[{lx'[t] == 10.(ly[t] - lx[t]),
               ly'[t] == lx[t](28. - lz[t]) - ly[t],
               lz'[t] == lx[t] ly[t] - (8./3.) lz[t],
               lx[0] == 1., ly[0] == 1., lz[0] == 1.},
              {lx, ly, lz}, {t, 0, 200}]];

ni = 10^4;
tbl = Table[{xx, Sin[xx]}, {xx, 0., 100., 100./(ni - 1)}];
ifn = Interpolation[tbl];
pts = RandomReal[{0, 100}, ni];
bench["Interpolation, 10^4 nodes, 10^4 evaluations", ifn /@ pts];
check["Interpolation at 3.3", Round[ifn[3.3], 1.*^-6]];

Print[""];
Print["re-run with MATHILDA_NO_AUTOCOMPILE=1 for the interpreted column"];
